Zero-length array allocation inspection (idea)

RusTit opened the pull request:
Text:
Reports on allocations of arrays with known lengths of zero. Since array lengths in Java are non-modifiable, it is almost always possible to share zero-length arrays, rather than repeatedly allocating new zero-length arrays. Such sharing may provide useful optimizations in program runtime or footprint. Note that this inspection does not report zero-length arrays allocated as static final fields, as it is assumed that those arrays are being used to implement array sharing.
And also I removed two methods. They are the same as those of the parent class.

RusTit committed:
AbstractMod.java

RusTit committed:
AlvearyLogicStimulator.java

RusTit committed:
AnalystPageFruit.java

RusTit committed:
BlockDesign.java

RusTit committed:
Botany.java remove override method. It is same with AbstractMod

RusTit committed:
ComponentTankContainer.java

RusTit committed:
ContainerCraftGUI.java

RusTit committed:
DefaultMachineInventory.java

RusTit committed:
DefaultTankContainer.java

RusTit committed:
Engineering.java

RusTit committed:
ErrorState.java

RusTit committed:
WindowGenesis.java

RusTit committed:
TransferRequest.java

RusTit committed:
WindowCompartment.java

RusTit committed:
MachineComponent.java, ManagerMachine.java, ItemSerumArray.java, RotatedOreIngredient.java

RusTit committed:
FlowerRoot.java - remove method override getRandomTemplate. It is same as in parent class.

RusTit committed:
Merge branch 'master-MC1.12' into memory_zero_Array

RusTit committed:
Refactoring and clean up

RusTit committed:
Small refactor

RusTit committed:
Small refactor

RusTit committed:
Small refactor

RusTit committed:
AlleleHelper.java - add final to vars in methods.

RusTit committed:
Refactoring BreedingSystem.java

RusTit commented:
@Nedelosk please review.

RusTit committed:
Gene.java

RusTit committed:
Code cleanup

RusTit committed:
Code cleanup

RusTit committed:
Code cleanup

RusTit committed:
Refactor to use EmptyHelper.EMPTY_STRING

RusTit committed:
Clean up fieldkit

RusTit committed:
Clean up WindowGenesis.java

RusTit committed:
Clean up ControlEnergyBar.java

RusTit committed:
Clean up ControlErrorState.java

RusTit committed:
Add final to vars

RusTit committed:
Small refactoring in ControlPlayerInventory.java

RusTit committed:
Refactor and localize ControlProgressBase.java

RusTit committed:
Add final to vars

RusTit committed:
Add Nullable annotation to label

RusTit committed:
Add final to vars

RusTit committed:
Add final to vars

RusTit committed:
Improve localise in ControlUser.java

RusTit committed:
Refactor to use EmptyHelper.EMPTY_STRING

RusTit committed:
Small refactoring in ContainerCraftGUI.java

RusTit committed:
Small refactoring in CustomSlot.java

RusTit committed:
Add final to vars GuiCraftGUI.java

RusTit committed:
Add final to vars Window.java

RusTit committed:
Add final to vars renderer

RusTit committed:
Add final to vars renderer

KorDum commented:
```java
Objects.equals(desc, EmptyHelper.EMPTY_STRING)
```
Hmm. Here is an code shorter:
```java
desc.equals(EmptyHelper.EMPTY_STRING)
```

But why do you write own constant for an empty string?
```java
import net.minecraft.util.StringUtils;
StringUtils.isNullOrEmpty(desc)
```
or
```java
import org.apache.commons.lang3.StringUtils;
desc.equals(StringUtils.EMPTY)
```

RusTit commented:
My bad. Java is not my main programming language. I was looking for a constant, but missed it. I'll correct it a bit later.

RusTit committed:
Refactor to use org.apache.commons.lang3.StringUtils.EMPTY

RusTit committed:
Fix crashing with NullPointerException in Tolerance.toString()

RusTit committed:
Add final to vars. Rename min and sec in ru_lang.

RusTit committed:
Refactor getWidget()
Remove:
```
//TODO: Why not use "this" ?
public IWidget getWidget() {
	return this;
}
```

RusTit committed:
Use Constants.EXTRA_BEES_MOD_ID in ExtraBeesIntegration.java

RusTit committed:
Remove this.setUnlocalizedName("genesis"); already invoked in super

RusTit committed:
Remove this.setUnlocalizedName("field_kit"); already invoked in super

RusTit committed:
Remove this.setUnlocalizedName(name); already invoked in super

RusTit committed:
Add final to vars.

RusTit committed:
Small refactoring in FluidContainerType.java

RusTit committed:
Add @Nullable annotation in FluidHandlerItemBinnie.java

RusTit committed:
Add final to vars.

RusTit committed:
Improve FluidContainerType.java and ItemFluidContainer.java localizing.

Nedelosk commented:
@RusTit Why are you adding final to every local var ?

temp1011 commented:
It can increase performance at runtime I believe but seems a bit controversial https://stackoverflow.com/questions/154314/when-should-one-use-final-for-method-parameters-and-local-variables

RusTit commented:
@Nedelosk
1. This does not break anything?
2. There may be a performance gain (or maybe not).
3. Should not be harm from this.
4. I have already started, I will finish in this PR only module core (so it is safer).

RusTit commented:
5.Makes code more standardized. Some files / methods already use the final in args, vars, for cycle.

Commit:
Add final and some refactoring in ManagerLiquid.java

Commit:
ManagerLiquid.java

Commit:
Add @Nullable annotation in ITankMachine.java

Commit:
Add final to vars in TileEntityMachineBase.java

Commit:
Add final to vars in ErrorState.java

Commit:
Remove unused anymore setOutputSides in BaseSlot.java

Commit:
Add final to vars and StringUtils.Empty for ""

Commit:
Refactor duplication code in ComponentInventorySlots.java, ComponentTankContainer.java

Commit:
Various final and annotations.

Commit:
Improve localize SlotValidator.java

Commit:
ControlSlot.java tooltip loc improve

Commit:
Small refactor TankSlot.java

Commit:
ComponentPowerReceptor.java - remove null checks from world var. But need deep testing.

Commit:
Various final, NBT keys const.

Commit:
Various final, NBT keys const.

Commit:
Compartment refactoring start.

RusTit commented on core/src/main/java/binnie/core/machines/storage/ComponentCompartmentInventory.java:
```diff
@@ -55,14 +58,15 @@ public void sendGuiNBTToClient(final Map<String, NBTTagCompound> nbt) {
-			final NBTTagList tags = nbt.getTagList("tabs", 10);
+	public void receiveGuiNBTOnClient(final EntityPlayer player, final String name, final NBTTagCompound nbt) {
+		// TODO: RusTit: I think here should be ACTION_COMP_CHANGE_TAB key, need testing.
+		if (name.equals(NBT_KEY_COMP_TABS)) {
```
To @Nedelosk @temp1011 
Need your advise.
As you can see, my commit do some refactoring.
Please take a look on methods: 
receiveGuiNBTOnClient and receiveGuiNBTOnServer
I think that in receiveGuiNBTOnClient code (original) `if (name.equals("comp-tabs")) {` is wrong.
Correct code should use `"comp-change-tab"` key (as in `receiveGuiNBTOnServer`).
It also used in `WindowCompartment.java` to provide action.

Nedelosk replied:
`receiveGuiNBTOnClient` only receives data from `sendGuiNBTToClient` so this should work fine. One thing I don't understand is why in `WindowCompartment` `sendClientAction` gets called with `"comp-tabs"`. https://github.com/ForestryMC/Binnie/blob/42435ea142bb28978625e461cd7b176af0d8979b/core/src/main/java/binnie/core/machines/storage/WindowCompartment.java#L109 if the only method that receives the data is `receiveGuiNBTOnServer` that only tests for `"comp-change-tab"`.

RusTit replied:
That is the problem.
Most likely it is a bug and a string the constant is specified incorrectly.

Commit:
SearchDialog.java code refactoring.

Commit:
Revert back checking for null.

Commit:
Improve SearchDialog.java localizing

RusTit committed:
Update ComponentPowerReceptor.java

RusTit committed:
Update ComponentPowerReceptor.java

Commit:
Fix border painting in SearchDialog

Commit:
Small clean up

RusTit commented:
Two latest commits fix border (lines near slots element) in compartment search window:
Fix:
![2018-04-11_13 27 32](https://user-images.githubusercontent.com/29846693/38597707-9d579c32-3d9c-11e8-9451-7a85de2f520c.png)
![2018-04-11_15 23 07](https://user-images.githubusercontent.com/29846693/38597708-9d9b70a6-3d9c-11e8-88aa-deee981446d4.png)

RusTit commented:
Search window has also bug with icon painting. ~~~According to my tests it is related to z-state of elements. In common situation all gui elements painted on 0 (zero) level.~~~ But Search window is created on top. Later then slots are repainted - it will produce bug.
Take a look on green ceramic block on first row (third from left). On this pic and pic above (last). If slot of compartment contain any item (block of wheat) it's mask will glitch on item icon in same position in search window. It also glitching in scrolling. @Nedelosk  @temp1011 what is best way to fix it? My idea with z-state (I think need only some changes in `ControlSlot` and `RenderUtil` to provide custom z-level painting)
![2018-04-11_15 23 11](https://user-images.githubusercontent.com/29846693/38597811-11514dd6-3d9d-11e8-89af-25ed0e156026.png)
![2018-04-11_15 23 16](https://user-images.githubusercontent.com/29846693/38597812-1185e046-3d9d-11e8-94f2-1802b93e3766.png)
![2018-04-11_15 23 24](https://user-images.githubusercontent.com/29846693/38597813-11bd75c4-3d9d-11e8-8d34-a87392c6619e.png)
![2018-04-11_15 23 28](https://user-images.githubusercontent.com/29846693/38597814-11f912e6-3d9d-11e8-8ffd-e535d5698ed1.png)

KorDum commented:
The key word `final` is usually used for constants in Java or readonly instance class level fields (initialize in constructor):
```java
private final int count;
public SomeClass(int count) {
    this.count = count;
}
```

Constants at compilation are inlined in bytecode in a number of cases.
It doesn't give any advantages in performance, it is a message to the programmer. By and large this clutters up the code and I would recommend removing everywhere (except for constants).

Java compiler itself optimizes the code when assembling to bytecode. JRE itself optimizes the bytecode when reading.
https://stackoverflow.com/questions/8354412/do-java-finals-help-the-compiler-create-more-efficient-bytecode

temp1011 commented:
Unfortunately I don't really know anything about GUIs. However, I think the point about `final` things is interesting.

Commit:
Remove unused empty Integer[] array anymore

mezz commented:
Most of these changes look good but I agree with @KorDum on `final`. Most of the existing `final` you see in the code are probably from the decompiler. It's not really that helpful in most cases. One thing I like to use it for is on `int` that are constant, since it is easy to read when you have some `int` that are obviously being mutated and some that are not.

RusTit commented:
Well, what to do with the refactoring already done. Should I remove the new finals.

mezz commented:
I think you should remove the new finals, sorry.

RusTit commented:
No problem, but it can take some time.

Commit:
Remove final from vars

Commit:
Merge remote-tracking branch 'main/HEAD' into memory_zero_Array

mezz commented on core/src/main/java/binnie/core/block/TileEntityMetadata.java:
```diff
@@ -40,8 +40,10 @@ public static ItemStack getItemStack(final Block block, final int damage) {
 		return itemStack;
 	}
 
+	private static final int MAX_ITEM_DAMAGE = 16387;
```
Any idea what the significance of this number is?
I thought it might be a power of 2 minus 1, but the nearest power of 2 is 2^14 (16384) which is also a weird number.

RusTit replied:
No idea, I was trying to find this value (binnie, forestry, forge) but no successed

temp1011 replied:
Max seems to be `OreDictionary.WILDCARD_VALUE` at 32767 so this doesn't even have the right number of bits.

mezz commented on core/src/main/java/binnie/core/genetics/Tolerance.java:
```diff
@@ -29,9 +33,19 @@ public static Tolerance get(final forestry.api.genetics.EnumTolerance tol) {
+	@Nullable
+	private static Pattern PATTERN;
+
+	private static Pattern GetPattern() {
```
Method names should start with a lowercase letter.

Actually I think this should just be created statically instead of using a singleton getter, since it is not very expensive to create and does not pull in lots of other classes.
```java
private static Pattern PATTERN = Pattern.compile("_", Pattern.LITERAL);
```

RusTit replied:
In this case it is not. At first version I do like you describe. But after mc run, it crushed with NullPointException in 
```
@Override
	public String toString() {
		return GetPattern().matcher(super.toString().toLowerCase(Locale.ENGLISH)).replaceAll(StringUtils.EMPTY);
}
```
Where is some issue with object creating (initial). At moment then toString invoked PATTERN will be null.

mezz replied:
Hm enums are weird, I guess they are being constructed before static. This is fine then.

mezz commented on core/src/main/java/binnie/core/gui/controls/button/ControlButton.java:
```diff
@@ -57,11 +57,13 @@ public String getText() {
 	@SideOnly(Side.CLIENT)
 	public void onRenderBackground(int guiWidth, int guiHeight) {
-		Object texture = CraftGUITexture.BUTTON_DISABLED;
+		Object texture;
```
One nice use of `final` is for cases like this. You can make `final Object texture;` and the static analyzer will complain "variable 'texture' might not have been initialized" if there is a case where it is not set.

RusTit replied:
I introduced the proposed changes, but it seems that the static analyzer worked correctly without them.

mezz commented on core/src/main/java/binnie/core/gui/database/PageBranchOverview.java:
```diff
@@ -42,15 +42,15 @@ public void onValueChanged(final IClassification branch) {
 		this.branchDescription.clear();
 		String desc = branch.getDescription();
-		if (desc == null || Objects.equals(desc, "") || desc.contains("for.")) {
+		if (desc == null || desc.length() == 0 || desc.contains("for.")) {
```
You can use `org.apache.commons.lang3.StringUtils#isEmpty` to check `desc`

mezz commented on core/src/main/java/binnie/core/gui/minecraft/control/ControlPlayerInventory.java:
```diff
@@ -41,10 +45,13 @@ public ControlPlayerInventory(final IWidget parent) {
+		for (int row = 0; row < rowCount; ++row) {
+			for (int column = 0; column < columnCount; ++column) {
+				final ControlSlot slot = new ControlSlot.Builder(this, column * 18, row * 18)
+                        .assign(InventoryType.PLAYER,column + row * columnCount);
```
this line is using spaces for indentation, the one above is tabs

mezz commented on core/src/main/java/binnie/core/gui/minecraft/control/ControlSlot.java:
```diff
@@ -193,11 +193,11 @@ public void getHelpTooltip(final Tooltip tooltip, ITooltipFlag tooltipFlag) {
 			tooltip.add(slot.getName());
 			if (tooltipFlag.isAdvanced()) {
-				Collection<EnumFacing> inputSides = slot.getInputSides();
+                Collection<EnumFacing> inputSides = slot.getInputSides();
```
this line indents with spaces

mezz commented on core/src/main/java/binnie/core/gui/minecraft/control/ControlSlot.java:
```diff
@@ -193,11 +193,11 @@ public void getHelpTooltip(final Tooltip tooltip, ITooltipFlag tooltipFlag) {
 					tooltip.add(TextFormatting.GRAY + I18N.localise(ModId.CORE, "gui.side.insert", MachineSide.asString(inputSides)));
 				}
-				Collection<EnumFacing> outputSides = slot.getOutputSides();
+                Collection<EnumFacing> outputSides = slot.getOutputSides();
```
this line indents with spaces

mezz commented on core/src/main/java/binnie/core/item/ItemMisc.java:
```diff
@@ -53,7 +52,7 @@ public ItemStack getStack(IItemMiscProvider type, int size) {
 	public void addInformation(ItemStack stack, @Nullable World worldIn, List<String> tooltip, ITooltipFlag flagIn) {
 		super.addInformation(stack, worldIn, tooltip, flagIn);
-		IItemMiscProvider item = this.getItem(stack.getItemDamage());
+        IItemMiscProvider item = this.getItem(stack.getItemDamage());
```
this line indents with spaces

mezz commented on core/src/main/java/binnie/core/item/ItemMisc.java:
```diff
@@ -67,7 +66,7 @@ public void registerModel(Item item, IModelManager manager) {
 	@Override
 	public String getItemStackDisplayName(final ItemStack stack) {
-		IItemMiscProvider item = this.getItem(stack.getItemDamage());
+        IItemMiscProvider item = this.getItem(stack.getItemDamage());
```
this line indents with spaces

mezz commented on core/src/main/java/binnie/core/liquid/ManagerLiquid.java:
```diff
@@ -39,20 +39,17 @@ public ManagerLiquid() {
+		if ((!FluidRegistry.registerFluid(bFluid)) || (FluidRegistry.addBucketForFluid(bFluid))) {
+			Log.error("Liquid registered incorrectly - {} ", fluid.getIdentifier());
+		}
+    }
```
this line indents with spaces

RusTit committed:
Method names should start with a lowercase letter.

RusTit committed:
Improve variable.

RusTit committed:
Change checking for string emptiness.

RusTit committed:
Improve intends

RusTit committed:
Improve localise in ControlLiquidTank.java

Commit:
Merge branch 'master-MC1.12' into memory_zero_Array

Commit:
Merge remote-tracking branch 'origin/memory_zero_Array' into memory_zero_Array

Commit:
Refactor tab slots array init.

Commit:
Extract createTab from lambda

Commit:
Refactoring compartment window.

Commit:
Code clean up

Commit:
Extract slide and localize

Commit:
Extract method createTabNameControl

Commit:
Extract method createTabIconControl

mezz reviewed (changes requested):
there are indentation issues in the latest commits

Commit:
Refactoring and localizing compartment

Commit:
indentation

mezz commented:
Please do not add more commits on to this PR, it needs to be reviewed and merged but that gets harder with more commits.
You can create a new PR for more changes.

RusTit commented:
Ok, but to make clear, where is still some bugs:
1. Item icons painting on second layer. (Some description is above).
2. Event key passing in ComponentCompartmentInventory.java  and WindowCompartment.java  (read my dialog with Nedelosk).

RusTit commented:
Item icon painting bug appears not only in the search window.
You can also see it in the slider (in which custom tab settings are set).

mezz commented:
If you have specific fixes for the changes made in this PR, I think that is fine.
I am just worried that this PR is going out of control, it seems to be fixing too many different things.

RusTit commented:
Not yet. I do not know how to fix a bug with rendering. I will fix the error with the key a little later with separate commits (will be last in this PR)

Commit:
Remove unused comment.

RusTit commented:
Apparently there is no bug with keys.
I did debugging by checking the passed string-keys through the events, in everything is ok.
P.S. but the bug with the renderer I can not solve myself.

KorDum commented:
Tabs indent or spaces?
